### PR TITLE
AEStreamParser: Fix E-AC-3 dependent stream passthrough (trac#17871)

### PR DESCRIPTION
**Pull request by anssih**
This fixes E-AC-3 dependent stream passthrough that got broken by FFmpeg 4.0 update: https://trac.kodi.tv/ticket/17871

There is one non-behavior-altering refactoring commit and then two fix commits, the latter of which is the bigger change with a commit message containing details about the issue(s) and the approach taken to fix them.

I've tested that the code now seems to work properly with standard AC-3, standard E-AC-3, and dependent-substream E-AC-3 regardless of whether the dependent frames are input as separate or combined.

### CODE REVIEW

**Review by fritsch — APPROVED** [review]

**Comment by fritsch** [comment]
Wait - the review button was only for:  https://github.com/xbmc/xbmc/pull/14604/commits/08054ee75680783a1c8ee3be8ab3f9a098fb5b30 need to check the other parts, too - I started with the smallest and most easy one.

**Comment by fritsch** [comment]
I runtime tested in the meantime. All my DD+ / AC3 testfiles still work as they should and those out of sync are in sync now. I realized that some of the 7.1 speaker-test samples are detected as

```
CAEStreamParser::TrySyncAC3 - E-AC3 stream detected (4 channels, 48000Hz)
```

but they are output fine. This was also the case before this PR, so not a problem, just an issue computing the right number of speakers from parsing acmod. Perhaps adding the number of channels of the dependend substream is missing here?

File: xbmc/cores/AudioEngine/Utils/AEStreamInfo.h [hunk 1]
```diff
@@ -87,13 +87,12 @@ class CAEStreamParser
...
   unsigned int m_dtsBlocks = 0;
   unsigned int m_fsize = 0;
-  unsigned int m_fsizeMain = 0;        /* used for EAC3 substreams */
   int m_substreams = 0;       /* used for TrueHD  */
   AVCRC m_crcTrueHD[1024];  /* TrueHD crc table */
 
   void GetPacket(uint8_t **buffer, unsigned int *bufferSize);
   unsigned int DetectType(uint8_t *data, unsigned int size);
-  bool TrySyncAC3(uint8_t *data, unsigned int size, bool resyncing);
+  bool TrySyncAC3(uint8_t *data, unsigned int size, bool resyncing, bool wantEAC3dependent = false);
```
Comment (fritsch) [inline xbmc/cores/AudioEngine/Utils/AEStreamInfo.h]: default parameters start to harm when they are changed somewhere. I would not use them, but rather being very explicit in each case.

**Review by fritsch — APPROVED** [review]

**Comment by fritsch** [comment]
https://github.com/xbmc/xbmc/pull/14604/commits/f068e40a439d68e3682334e67d9bd7607eba6b5d is review with a minor nitpick about default parameter used.

**Comment by fritsch** [comment]
All find from my side, still a beta time to fix regressions. Want to here @FernetMenta oppionon also, please.

**Comment by anssih** [comment]
> I realized that some of the 7.1 speaker-test samples are detected as
> 
> ```
> CAEStreamParser::TrySyncAC3 - E-AC3 stream detected (4 channels, 48000Hz)
> ```
> but they are output fine. This was also the case before this PR, so not a problem, just an issue computing the right number of speakers from parsing acmod. Perhaps adding the number of channels of the dependend substream is missing here?

Noticed that as well. It is because the channel information is only parsed from the dependent substream.  We would need to merge (or something non-trivial, didn't look into it) the channel maps from both streams to get a correct count.

**Pull request merged**